wgengine/magicsock/derp.go

```diff
@@ -158,13 +158,17 @@ func (c *Conn) maybeSetNearestDERP(report *netcheck.Report) (preferredDERP int)
 	//
 	// For tests, always assume we're connected to control unless we're
 	// explicitly testing this behaviour.
+	//
+	// Despite the above behaviour, ensure that we set the nearest DERP if
+	// we don't currently have one set; any DERP server is better than
+	// none, even if not connected to control.
 	var connectedToControl bool
 	if testenv.InTest() && !checkControlHealthDuringNearestDERPInTests {
 		connectedToControl = true
 	} else {
 		connectedToControl = c.health.GetInPollNetMap()
 	}
-	if !connectedToControl {
+	if !connectedToControl && c.myDerp != 0 {
```

Comment:
maybe also check the incoming one is not 0?

Reply:
oh also this is a race. `c.mu` needs to be held

Reply:
I think we don't want to do that, since we do want to fall through to picking a deterministic-but-random one; added a test to confirm this works.

Reply:
Fixed race; thanks for catching that.